ykbh/src/lib.rs

```diff
@@ -0,0 +1,310 @@
+                Statement::Unimplemented(_) | Statement::Debug(_) => { todo!("Unimplemented") },
+                Statement::Cast(..) => todo!("cast"),
+                Statement::Call(..) | Statement::StorageDead(_) => unreachable!()
+            }
+        }
+
+        match &body.blocks[0].term {
+            Terminator::Call {
+                operand: _op,
+                args: _args,
+                destination: _dest
+            } => {
+                todo!()
+            }
+            Terminator::Return => {
+            }
+            t => todo!("{}", t)
+        }
+    }
+
+    /// Get the Object referenced by a Local.
+    fn get_var(&self, local: &Local) -> &Option<Object> {
+        &self.vars[local.0 as usize]
+    }
+
+    /// Implements the Store statement.
+    fn store(&mut self, dest: &IPlace, src: &IPlace) {
+        self.store_obj(dest, self.to_obj(src));
+    }
+
+    /// Write an Object to an IPlace. This either creates a new Object if the IPlace is a Val, or
+    /// writes the data to some pointer, if the IPlace is an Indirect.
+    fn store_obj(&mut self, dest: &IPlace, src: Object) {
+        match dest {
+            IPlace::Val { local, off, ty: _ty } => {
+                // Store into a local.
+                if *off == 0 {
+                    self.vars[local.0 as usize] = Some(src);
+                } else {
+                    todo!()
+                }
+            }
+            IPlace::Indirect { ptr, off, ty: _ty } => {
+                // Write to a pointer.
+                if let Some(obj) = self.get_var(&ptr.local) {
+                    let dstptr = obj.to_usize() + ptr.off as usize + *off as usize;
+                    self.store_raw(dstptr, src);
+                }
+            }
+            _ => unreachable!()
+        };
+    }
+
+    fn store_raw(&self, dest: usize, src: Object) {
+        unsafe {
+            std::ptr::copy(src.data.as_ptr(), dest as *mut u8, src.data.len());
+        }
+    }
+
+    fn to_obj(&self, src: &IPlace) -> Object {
+        match src {
+            IPlace::Val { local, off, ty } => {
+                if *off == 0 {
+                    // If the offset is 0 we can just copy the object, e.g. $1 = $2.
+                    if let Some(obj) = self.get_var(local) {
+                        obj.clone()
+                    } else {
+                        unreachable!()
+                    }
+                } else {
+                    // Copy data from an offset, e.g. $1 = $2+2.
+                    if let Some(obj) = self.get_var(local) {
+                        let ptr = obj.data.as_ptr() as usize + *off as usize;
+                        let size = SIR.ty(ty).size();
+                        Object::from_ptr(ptr, size as usize)
+                    } else {
+                        unreachable!()
+                    }
+                }
+            }
+            IPlace::Indirect { ptr, off, ty } => {
+                // FIXME copy only depending on type?
```

Comment:
Can you expand on this FIXME?

Reply:
I'm currently unsure if there are cases where we don't copy the thing this points to, but just return a reference to it instead. I think it depends on the type of the thing, e.g. a tuple is copied, a vec isn't.

Reply:
A `Vec` value is copied. It is just that the storage pointed to by the internal pointer is not copied.

Reply:
I think at the SIR level every store is a copy.